File: java/leetcode/problem40/Solution1.java
package leetcode.problem40;
import core.Solution;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.*;

public class Solution1 implements Solution {
    int[] cands;
    List<List<Integer>> ans;
    LinkedList<Integer> cur;
    
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        this.cands = candidates;
        this.ans = new ArrayList<>();
        this.cur = new LinkedList<>();
        dfs(0, target);
        return ans;
    }
    void dfs(int i, int target) {
        if (target == 0) {
            ans.add(new ArrayList<>(cur));
            return;
        }
        if (i >= cands.length) return;
        for (int j = i; j < cands.length; ++j) {
            if (j > i && cands[j] == cands[j - 1]) continue;
            if (target < cands[j]) break;
            cur.add(cands[j]);
            dfs(j + 1, target - cands[j]);
            cur.removeLast();
        }
    }
    
    @Override
    public String describe() {
        return "Backtracking";
    }
    
    @Override
    public URL link() throws MalformedURLException {
        return new URL("https://leetcode.cn/problems/combination-sum-ii/");
    }
}

File: go/leetcode/problem56/solution1.go
package problem56

import "sort"

// Better solution!
type Solution1 struct{}

func (self *Solution1) Merge(intervals [][]int) [][]int {
	sort.Slice(intervals, func(i, j int) bool {
		if intervals[i][1] == intervals[j][1] {
			return intervals[i][0] < intervals[j][0]
		}
		return intervals[i][1] < intervals[j][1]
	})
	return self.dfs(intervals)
}

func (self *Solution1) dfs(xs [][]int) [][]int {
	if len(xs) <= 1 {
		return xs
	}
	x := xs[0]
	rest := self.dfs(xs[1:])
	n := len(rest)
	y := rest[n-1]
	if x[1] < y[0] {
		rest = append(rest, x)
		return rest
	}
	if x[0] >= y[0] {
		return rest
	}
	rest = rest[:n-1]
	rest = append(rest, []int{x[0], y[1]})
	return rest
}
